client/app/pages/queries/QuerySource.jsx

```diff
@@ -1,19 +1,174 @@
+    });
+}
+
+function chooseDataSourceId(dataSourceIds, availableDataSources) {
+  dataSourceIds = map(dataSourceIds, v => parseInt(v, 10));
+  availableDataSources = map(availableDataSources, ds => ds.id);
+  return find(dataSourceIds, id => includes(availableDataSources, id)) || null;
+}
+
+function QuerySource(props) {
+  const [query, setQuery] = useState(props.query);
+  const [allDataSources, setAllDataSources] = useState([]);
+  const [dataSourcesLoaded, setDataSourcesLoaded] = useState(false);
+  const dataSources = useMemo(() => filter(allDataSources, ds => !ds.view_only || ds.id === query.data_source_id), [
+    allDataSources,
+    query.data_source_id,
+  ]);
+  const dataSource = useMemo(() => find(dataSources, { id: query.data_source_id }) || null, [query, dataSources]);
+  const [schema, setSchema] = useState([]);
+  const refreshSchemaTokenRef = useRef(null);
+
+  useEffect(() => {
+    let cancelDataSourceLoading = false;
+    DataSource.query().$promise.then(data => {
+      if (!cancelDataSourceLoading) {
+        setDataSourcesLoaded(true);
+        setAllDataSources(data);
+      }
+    });
+
+    return () => {
+      // cancel pending operations
+      cancelDataSourceLoading = true;
+      refreshSchemaTokenRef.current = null;
+    };
+  }, []);
+
+  const reloadSchema = useCallback(
+    (refresh = undefined) => {
+      const refreshToken = Math.random()
+        .toString(36)
+        .substr(2);
+      refreshSchemaTokenRef.current = refreshToken;
+      getSchema(dataSource, refresh).then(data => {
+        if (refreshSchemaTokenRef.current === refreshToken) {
+          setSchema(data);
+        }
+      });
+    },
+    [dataSource]
+  );
+
+  useEffect(() => {
+    reloadSchema();
+  }, [reloadSchema]);
+
+  useEffect(() => {
+    recordEvent("view_source", "query", query.id);
+  }, [query.id]);
+
+  const handleDataSourceChange = useCallback(
+    dataSourceId => {
+      if (dataSourceId) {
+        try {
+          localStorage.setItem("lastSelectedDataSourceId", dataSourceId);
+        } catch (e) {
+          // `localStorage.setItem` may throw exception if there are no enough space - in this case it could be ignored
+        }
+      }
+      if (query.data_source_id !== dataSourceId) {
+        recordEvent("update_data_source", "query", query.id, { dataSourceId });
+        const newQuery = clone(query);
+        newQuery.data_source_id = dataSourceId;
+        newQuery.latest_query_data = null;
+        newQuery.latest_query_data_id = null;
+        setQuery(newQuery);
+        // TODO: Save if not new
+      }
+    },
+    [query]
+  );
+
+  useEffect(() => {
+    // choose data source id for new queries
+    if (dataSourcesLoaded && query.isNew()) {
+      const firstDataSourceId = dataSources.length > 0 ? dataSources[0].id : null;
+      handleDataSourceChange(
+        chooseDataSourceId(
+          [query.data_source_id, localStorage.getItem("lastSelectedDataSourceId"), firstDataSourceId],
```

Comment:
Any practical difference using the getters/setters?

Reply:
[MDN article](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API) recommends to use getter/setters, and I found that arguments strong enough. So I took the opportunity and replaced props with getters/setters.